### PR TITLE
Helsinki-Testbed

### PR DESCRIPTION
**Pull request by tomtom7979**
* Popup.tsx --> Code entity that calculates Apartment Buildings energy efficiency
* layerConf.ts --> Demolished buildings color to black

### CODE REVIEW

File: legacy/webpack.config.babel.js [hunk 1]
```diff
@@ -61,7 +61,7 @@ const appConfig = {
   devtool: 'source-map',
   entry: ['./src/Entry.tsx'],
   resolve: {
-    extensions: ['.webpack-loader.js', '.web-loader.js', '.loader.js', '.js', '.jsx', '.ts', '.tsx'],
+    extensions: ['.webpack-loader.js', '.web-loader.js', '.loader.js', '.js', '.jsx', '.ts', '.tsx', '.json'],
```
Comment (xbexbex) [inline legacy/webpack.config.babel.js:64]: What is this used for? I don't think you added json files anywhere.

File: package.json [hunk 2]
```diff
@@ -104,6 +105,7 @@
     "eslint-webpack-plugin": "^3.2.0",
     "git-revision-webpack-plugin": "^5.0.0",
     "html-webpack-plugin": "^5.5.0",
+    "json-loader": "^0.5.7",
```
Comment (xbexbex) [inline package.json:108]: Same as above. Please remove the import if it not used.

File: package.json [hunk 3]
```diff
@@ -72,6 +72,7 @@
   "devDependencies": {
     "@babel/core": "^7.20.2",
     "@babel/eslint-parser": "^7.19.1",
+    "@babel/plugin-proposal-object-rest-spread": "^7.19.4",
```
Comment (xbexbex) [inline package.json:75]: What is this used for? Please remove the import if it not used.

File: src/components/Map/Layers/Buildings/HelsinkiBuildings/Popup.tsx [hunk 4]
```diff
@@ -2,6 +2,51 @@ import React from 'react'
 import Feature from 'ol/Feature'
 import { Table, TableBody, TableCell, TableRow } from '@mui/material'
 import _ from 'lodash'
+// Variables
+var heatings
```
Comment (xbexbex) [inline src/components/Map/Layers/Buildings/HelsinkiBuildings/Popup.tsx:6]: Could you change all the "var" declarations in the code to "const" if they don't change? And to "let" if they are redeclared.

File: src/components/Map/Layers/Buildings/HelsinkiBuildings/Popup.tsx [hunk 5]
```diff
@@ -2,6 +2,51 @@ import React from 'react'
...
+  return teconss
+}
+// Emission factor function [kgCO2/kWh]
+function emissionFactor(tecc, empdpn) {  
+  emisfactord = Math.round((tecc * empdpn)/1000)*1000    
+  emisfactords = emisfactord.toString().slice(0, -3);  
+  emisfactords = `${emisfactords} 000`
+  return emisfactords;
+}
+// Estimated annual energy consumption [kWh/m3,a]
```
Comment (xbexbex) [inline src/components/Map/Layers/Buildings/HelsinkiBuildings/Popup.tsx:35]: It seems a bit redundant to transform the JSON-string to javascript object every time the code is run. Could you pre-convert the json to object? You can use the console on any webpage to run this code and just paste the resulting object here.

Also, please split large constant objects to a "constants.ts"-file and export from there. See [https://github.com/AvoinOrg/climate-map/blob/ol/src/components/Map/Layers/Forests/FinlandMatureForests/constants.ts](https://github.com/AvoinOrg/climate-map/blob/ol/src/components/Map/Layers/Forests/FinlandMatureForests/constants.ts)

**Review by xbexbex — CHANGES_REQUESTED** [review]
Very good! Just a couple of minor nitpicks :)

**Comment by tomtom7979** [comment]
Hi, I made changes and pushed the changes to ol-ti branch.

**Pull request merged**